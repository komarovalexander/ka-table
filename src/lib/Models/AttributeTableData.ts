import { HTMLAttributes } from 'react';

import { ChildProps, DispatchFunc } from '../types';

export class AttributeTableData {
  public baseFunc!: any;
  public childElementAttributes!: HTMLAttributes<HTMLElement>;
  public childProps!: ChildProps;
  public dispatch!: DispatchFunc;
}
