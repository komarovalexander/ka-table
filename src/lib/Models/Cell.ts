/**
 * Describes the position of cell in table
 */
export class Cell {
  /*
  * Specifies the field of data
  */
  public columnKey!: string;

  /*
  * Specifies the value of row's key where cell is placed
  */
  public rowKey!: any;
}
