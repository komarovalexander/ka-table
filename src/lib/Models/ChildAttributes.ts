
import { ITableAllProps } from '../';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { ChildAttributesItem } from '../types';

export class ChildAttributes {
  public cell?: ChildAttributesItem<ICellContentProps>;
  public table?: ChildAttributesItem<ITableAllProps>;
}
