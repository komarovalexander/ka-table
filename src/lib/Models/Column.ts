import { DataType, SortDirection } from '../enums';
import {
  CellFunc, EditorFunc, FilterRowFunc, FormatFunc, HeaderCellFunc, SearchFunc, ValidationFunc,
} from '../types';

/**
 * Describes column of table its look and behaviour
 */
export class Column {
  public cell?: CellFunc;
  public dataType?: DataType;
  public editor?: EditorFunc;
  public filterRowCell?: FilterRowFunc;
  public filterRowValue?: any;
  public filterRowOperator?: any;
  public field?: string;
  public format?: FormatFunc;
  public headCell?: HeaderCellFunc;
  public isEditable?: boolean;
  public key!: string;
  public search?: SearchFunc;
  public sortDirection?: SortDirection;
  public style?: React.CSSProperties;
  public title?: string;
  public validation?: ValidationFunc;
}
