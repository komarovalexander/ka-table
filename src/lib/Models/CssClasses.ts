export class CssClasses {
  public table?: string = 'ka-table';
  public tbody?: string = 'ka-tbody';
  public thead?: string = 'ka-thead';
  public theadRow?: string = 'ka-tr ka-thead-row';
  public theadCell?: string = 'ka-thead-cell';
  public groupRow?: string = 'ka-tr ka-group-row';
  public row?: string = 'ka-tr ka-row';
  public rowSelected?: string = 'ka-row-selected';
  public cell?: string = 'ka-cell';

  public iconGroupArrowCollapsed?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-collapsed';
  public iconGroupArrowExpanded?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-expanded';
  public iconClose?: string = 'ka-icon ka-icon-close';
  public iconSortArrowDown?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-down';
  public iconSortArrowUp?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-up';

  public checkbox ?: string = 'ka-input';
  public dateInput ?: string = 'ka-input';
  public numberInput ?: string = 'ka-input';
  public textInput ?: string = 'ka-input';
}
