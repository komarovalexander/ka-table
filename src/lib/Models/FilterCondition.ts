/* Filter condition for specific field */
export class FilterCondition {
  public field!: string;
  public operator!: string;
  public value!: any;
}
