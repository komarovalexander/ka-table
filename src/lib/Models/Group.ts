
export class Group {
  public columnKey!: string;
}
