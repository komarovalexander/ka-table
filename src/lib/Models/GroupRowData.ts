export class GroupRowData {
  public groupMark!: object;
  public key!: any[];
  public value!: any;
}
