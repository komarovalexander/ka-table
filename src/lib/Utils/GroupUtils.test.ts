import { Group } from '../Models/Group';
import { GroupRowData } from '../Models/GroupRowData';
import {
  convertToFlat, getExpandedGroups, getGroupedStructure, getGroupMark, groupBy, groupClick,
} from './GroupUtils';

describe('GroupUtils', () => {
  const columnMap = { country: 'country', type: 'type' };
  const data = [
    { type: 'Cat', name: 'Kas', country: 'Czech Republic' },
    { type: 'Dog', name: 'Rex', country: 'Montenegro' },
    { type: 'Cat', name: 'Simba', country: 'France' },
    { type: 'Dog', name: 'Beethoven', country: 'Czech Republic' },
    { type: 'Cat', name: 'Hash', country: 'Czech Republic' },
  ];
  const groups: Group[] = [{ columnKey: 'country' }, { columnKey: 'type' }];

  describe('convertToFlat', () => {
    it('simple', () => {
      const mappedData = new Map([['France', [{
        country: 'France',
        name: 'Simba',
        type: 'Cat',
      }]]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });

    it('inner group', () => {
      const mappedData = new Map([['France',
        new Map([['Cat', [{
          country: 'France',
          name: 'Simba',
          type: 'Cat',
        }]]]),
      ]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });

    it('complex structure', () => {
      const mappedData = new Map([['France',
        new Map([
          ['Cat', [{
            country: 'France',
            name: 'Simba',
            type: 'Cat',
          }, {
            country: 'France',
            name: 'Kas',
            type: 'Cat',
          }]],
          ['Dog', [{
            country: 'France',
            name: 'Slimu',
            type: 'Dog',
          }, {
            country: 'France',
            name: 'Rex',
            type: 'Dog',
          }]],
        ]),
      ],
      ['Brazil',
        new Map([
          ['Cat', [{
            country: 'Brazil',
            name: 'Simba',
            type: 'Cat',
          }, {
            country: 'Brazil',
            name: 'Kas',
            type: 'Cat',
          }]],
          ['Dog', [{
            country: 'Brazil',
            name: 'Slimu',
            type: 'Dog',
          }, {
            country: 'Brazil',
            name: 'Rex',
            type: 'Dog',
          }]],
        ]),
      ]]);
      const result = convertToFlat(mappedData);
      expect(result).toMatchSnapshot();
    });
  });

  it('getExpandedGroups', () => {
    const groupRowData: GroupRowData = { key: ['dog'], value: 'Rex', groupMark: getGroupMark() };
    const groupedData: any[] = [{}, groupRowData, {}];
    const result = getExpandedGroups(groupedData);
    expect(result).toEqual([['dog']]);
  });

  it('groupClick add', () => {
    const groupsExpanded: any[][] = [['cat']];
    const groupRowData: GroupRowData = { key: ['dog'], value: 'Rex', groupMark: {} };
    const onOptionChange = jest.fn((x) => {});
    groupClick(groupsExpanded, groupRowData, onOptionChange);
    expect(onOptionChange.mock.calls.length).toBe(1);
    expect(onOptionChange.mock.calls[0]).toEqual([{groupsExpanded: [['cat'], ['dog']]}]);
  });

  it('groupClick remove', () => {
    const groupsExpanded: any[][] = [['cat']];
    const groupRowData: GroupRowData = { key: ['cat'], value: 'Tom', groupMark: {} };
    const onOptionChange = jest.fn((x) => {});
    groupClick(groupsExpanded, groupRowData, onOptionChange);
    expect(onOptionChange.mock.calls.length).toBe(1);
    expect(onOptionChange.mock.calls[0]).toEqual([{groupsExpanded: []}]);
  });

  it('groupBy', () => {
    const result = groupBy(data, (item: any) => item.type);
    expect(result).toMatchSnapshot();
  });

  it('groupBy empty', () => {
    const result = groupBy(data, (item: any) => item.type, true);
    expect(result).toMatchSnapshot();
  });

  describe('getGroupedStructure', () => {
    it('basic', () => {
      const result = getGroupedStructure(data, groups, columnMap);
      expect(result).toMatchSnapshot();
    });

    it('expanded root', () => {
      const result = getGroupedStructure(data, groups, columnMap, 0, [['Czech Republic']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded second', () => {
      const result = getGroupedStructure(data, groups, columnMap, 0, [['Czech Republic'], ['Czech Republic', 'Cat']]);
      expect(result).toMatchSnapshot();
    });
    it('should not expand second', () => {
      const result = getGroupedStructure(data, groups, columnMap, 0, [['Czech Republic', 'Cat']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded couple', () => {
      const result = getGroupedStructure(data, groups, columnMap, 0,
        [['Czech Republic'], ['Czech Republic', 'Cat'], ['Montenegro']]);
      expect(result).toMatchSnapshot();
    });
    it('expanded couple (skip Czech Republic)', () => {
      const result = getGroupedStructure(data, groups, columnMap, 0, [['Czech Republic', 'Cat'], ['Montenegro']]);
      expect(result).toMatchSnapshot();
    });
  });
});
