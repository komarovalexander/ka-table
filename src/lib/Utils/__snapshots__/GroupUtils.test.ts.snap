// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupUtils convertToFlat complex structure 1`] = `
Array [
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
  Object {
    "country": "France",
    "name": "Kas",
    "type": "Cat",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
      "Dog",
    ],
    "value": "Dog",
  },
  Object {
    "country": "France",
    "name": "Slimu",
    "type": "Dog",
  },
  Object {
    "country": "France",
    "name": "Rex",
    "type": "Dog",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "Brazil",
    ],
    "value": "Brazil",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "Brazil",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "Brazil",
    "name": "Simba",
    "type": "Cat",
  },
  Object {
    "country": "Brazil",
    "name": "Kas",
    "type": "Cat",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "Brazil",
      "Dog",
    ],
    "value": "Dog",
  },
  Object {
    "country": "Brazil",
    "name": "Slimu",
    "type": "Dog",
  },
  Object {
    "country": "Brazil",
    "name": "Rex",
    "type": "Dog",
  },
]
`;

exports[`GroupUtils convertToFlat inner group 1`] = `
Array [
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
      "Cat",
    ],
    "value": "Cat",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
]
`;

exports[`GroupUtils convertToFlat simple 1`] = `
Array [
  Object {
    "groupMark": Object {},
    "key": Array [
      "France",
    ],
    "value": "France",
  },
  Object {
    "country": "France",
    "name": "Simba",
    "type": "Cat",
  },
]
`;

exports[`GroupUtils getGroupedStructure basic 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Beethoven",
        "type": "Dog",
      },
    ],
  },
  "Montenegro" => Map {
    "Dog" => Array [
      Object {
        "country": "Montenegro",
        "name": "Rex",
        "type": "Dog",
      },
    ],
  },
  "France" => Map {
    "Cat" => Array [
      Object {
        "country": "France",
        "name": "Simba",
        "type": "Cat",
      },
    ],
  },
}
`;

exports[`GroupUtils getGroupedStructure expanded couple (skip Czech Republic) 1`] = `
Map {
  "Czech Republic" => Array [],
  "Montenegro" => Map {
    "Dog" => Array [],
  },
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded couple 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [],
  },
  "Montenegro" => Map {
    "Dog" => Array [],
  },
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded root 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [],
    "Dog" => Array [],
  },
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure expanded second 1`] = `
Map {
  "Czech Republic" => Map {
    "Cat" => Array [
      Object {
        "country": "Czech Republic",
        "name": "Kas",
        "type": "Cat",
      },
      Object {
        "country": "Czech Republic",
        "name": "Hash",
        "type": "Cat",
      },
    ],
    "Dog" => Array [],
  },
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils getGroupedStructure should not expand second 1`] = `
Map {
  "Czech Republic" => Array [],
  "Montenegro" => Array [],
  "France" => Array [],
}
`;

exports[`GroupUtils groupBy 1`] = `
Map {
  "Cat" => Array [
    Object {
      "country": "Czech Republic",
      "name": "Kas",
      "type": "Cat",
    },
    Object {
      "country": "France",
      "name": "Simba",
      "type": "Cat",
    },
    Object {
      "country": "Czech Republic",
      "name": "Hash",
      "type": "Cat",
    },
  ],
  "Dog" => Array [
    Object {
      "country": "Montenegro",
      "name": "Rex",
      "type": "Dog",
    },
    Object {
      "country": "Czech Republic",
      "name": "Beethoven",
      "type": "Dog",
    },
  ],
}
`;

exports[`GroupUtils groupBy empty 1`] = `
Map {
  "Cat" => Array [],
  "Dog" => Array [],
}
`;
